keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,295 @@
+
+Defaulting is implemented for most native Kubernetes API types and plays a crucial role for API compatibility when adding new fields. CustomResources do not support this natively. 
+
+Mutating admission webhooks can be used to partially replicate defaulting for incoming request payloads. But mutating admission webhooks do not run when reading from etcd.
+
+This KEP is about adding support for specifying default values for fields via OpenAPI v3 validation schemas in the CRD manifest. OpenAPI v3 has native support for a `default` field with arbitrary JSON values, for example: 
+
+```yaml
+properties:
+  foo:
+    type: string
+    default: "abc" 
+```
+
+This KEP proposes to apply these default values during deserialization, in the same way as native resources do. We assume _structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002).
+
+## Motivation
+
+* By far most native Golang based resources implement defaulting. CRD do not allow that, leading to unnatural, not Kubernetes-like APIs. This is bad for the ecosystem.
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequat as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002) makes defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults. Native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](https://github.com/kubernetes/enhancements/pull/1002)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+  * undefined (`_, ok := obj[field]; !ok`)
+  * `nil` if the field not nullable.
+
+We do this in the serializer just after the binary payload has been unmarshalled into an `map[string]interface{}`, pruning of [KEP: Pruning for CustomResources](https://github.com/kubernetes/enhancements/pull/709) was done, compare the yellow boxes in the following figure:
```

Comment:
Can we do it in a way that is extremely similar to native-type, by plugging in some `Defaulter` and do it about at the same place?

Reply:
yes, exactly this is the plan.

My original go-openapi (non-structural) implementation (https://github.com/kubernetes/kubernetes/pull/63604) was already near to this goal, doing defaulting during decoding (because pruning and defaulting was connected; not so anymore with structural schemas). So now we can [pass a real defaulter to the codec](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49), i.e. https://github.com/kubernetes/kubernetes/pull/63604/commits/09f02ba8f971613b47ad4fd7d73ca2f509524ed0#diff-2f720a7e61a4ddfcfe9a9e353be8784bR751 will not pass a noop variant, but the real thing.

The result is the big advantage compared to admission webhooks: we can do defaulting exactly where native resource do it, also in the storage layer where admission webhooks are not called.

Reply:
Clarified this in the text.